src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,714 @@
+/*!
+ * Copyright (c) 2018 by Contributors
+ *
+ * \file partial_eval.cc
+ *
+ * \brief Perform known computation in compile time.
+ *
+ * The partial evaluator try to do computation at compile time,
+ * so it can generate code that do less work.
+ * Additionally, it might open more chance for further optimization,
+ * since the high level, structural part of the code (closure, reference, control flow)
+ * might get partially evaluated away, and the subsequent optimization (for example, kernel fusion)
+ * can reason across those structural code as it got removed.
+ * In the extreme case, partial evaluation can even turn the whole program
+ * into pure first order computation with no control.
+ * In such a case, we can compile the whole computation onto SIMD Instruction/GPU/FPGA,
+ * and get insane speedup.
+ *
+ * It works by making the following modifications to the standard relay interpreter:
+ *
+ * 0: The values become partially static value.
+ * Since we cannot know the value of every term at compile time,
+ * Term might get partially evaluated to 'Unknown Value'.
+ * Every partially static value is, hence,
+ * a static fragment that might not be there (partially static),
+ * and a dynamic fragment that is semantically equivalent to the original term,
+ * so the unknown part will be computed at runtime, using the dynamic fragment.
+ *
+ * 1: The interpreter get paired with a LetList, which preserves ANF for the generated code.
+ * More specifically, we require that all dynamic is an atom.
+ * This avoid code duplication (which is both inefficient and incorrect) as atom has constant size,
+ * and allow us to not handle capture-avoidance substitution (as atom has no binder).
+ *
+ * 2: The store get reified.
+ * Instead of Reference having mutable field, Reference only has an unique identifier.
+ * There will be a mutable mapping of id to partially static value, called the store.
+ * This allow us to rollback the store:
+ * when a path may or may not get executed, we copy the store, recurse with a copy,
+ * and when we are done, we continue with the old store, so those effect 'do not happened'.
+ * We do this in if else, pattern matching, and in function,
+ * as, when we see a function, we partially evaluate it with all the argument as dynamic,
+ * to generate efficient dynamic for that function.
+ *
+ * 3: The code generated will reuse binding (although they are not shadowed),
```

Comment:
```suggestion
 * 3: The generated code reuses bindings (although they are not shadowed),
```

to follow the verb structure of the previous two numbered summaries